# Multi-task recommenders

**Learning Objectives**
  1. Training a model which focuses on ratings.
  2. Training a model which focuses on retrieval.
  3. Training a joint model that assigns positive weights to both ratings & retrieval models.


## Introduction
In the basic retrieval notebook we built a retrieval system using movie watches as positive interaction signals.

In many applications, however, there are multiple rich sources of feedback to draw upon. For example, an e-commerce site may record user visits to product pages (abundant, but relatively low signal), image clicks, adding to cart, and, finally, purchases. It may even record post-purchase signals such as reviews and returns.

Integrating all these different forms of feedback is critical to building systems that users love to use, and that do not optimize for any one metric at the expense of overall performance.

In addition, building a joint model for multiple tasks may produce better results than building a number of task-specific models. This is especially true where some data is abundant (for example, clicks), and some data is sparse (purchases, returns, manual reviews). In those scenarios, a joint model may be able to use representations learned from the abundant task to improve its predictions on the sparse task via a phenomenon known as [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning). For example, [this paper](https://openreview.net/pdf?id=SJxPVcSonN) shows that a model predicting explicit user ratings from sparse user surveys can be substantially improved by adding an auxiliary task that uses abundant click log data.

In this jupyter notebook, we are going to build a multi-objective recommender for Movielens, using both implicit (movie watches) and explicit signals (ratings).

Each learning objective will correspond to a __#TODO__ in the [student lab notebook](https://github.com/GoogleCloudPlatform/training-data-analyst/blob/master/courses/machine_learning/deepdive2/recommendation_systems/labs/multitask.ipynb) -- try to complete that notebook first before reviewing this solution notebook.


## Imports


Let's first get our imports out of the way.


In [2]:
# Installing the necessary libraries.
!pip install -q tensorflow-recommenders
!pip install -q --upgrade tensorflow-datasets

**NOTE: Please ignore any incompatibility warnings and errors and re-run the above cell before proceeding.**


In [3]:
# Importing the necessary modules
import os
import pprint
import tempfile

from typing import Dict, Text

import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds

In [4]:
import tensorflow_recommenders as tfrs

## Preparing the dataset

We're going to use the Movielens 100K dataset.

In [5]:
ratings = tfds.load('movielens/100k-ratings', split="train")
movies = tfds.load('movielens/100k-movies', split="train")

# Select the basic features.
ratings = ratings.map(lambda x: {
    "movie_title": x["movie_title"],
    "user_id": x["user_id"],
    "user_rating": x["user_rating"],
})
movies = movies.map(lambda x: x["movie_title"])

And repeat our preparations for building vocabularies and splitting the data into a train and a test set:

In [6]:
# Randomly shuffle data and split between train and test.
tf.random.set_seed(42)
shuffled = ratings.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

train = shuffled.take(80_000)
test = shuffled.skip(80_000).take(20_000)

movie_titles = movies.batch(1_000)
user_ids = ratings.batch(1_000_000).map(lambda x: x["user_id"])

unique_movie_titles = np.unique(np.concatenate(list(movie_titles)))
unique_user_ids = np.unique(np.concatenate(list(user_ids)))

## A multi-task model

There are two critical parts to multi-task recommenders:

1. They optimize for two or more objectives, and so have two or more losses.
2. They share variables between the tasks, allowing for transfer learning.

In this jupyter notebook, we will define our models as before, but instead of having  a single task, we will have two tasks: one that predicts ratings, and one that predicts movie watches.

The user and movie models are as before:

```python
user_model = tf.keras.Sequential([
  tf.keras.layers.experimental.preprocessing.StringLookup(
      vocabulary=unique_user_ids, mask_token=None),
  # We add 1 to account for the unknown token.
  tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
])

movie_model = tf.keras.Sequential([
  tf.keras.layers.experimental.preprocessing.StringLookup(
      vocabulary=unique_movie_titles, mask_token=None),
  tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
])
```

However, now we will have two tasks. The first is the rating task:

```python
tfrs.tasks.Ranking(
    loss=tf.keras.losses.MeanSquaredError(),
    metrics=[tf.keras.metrics.RootMeanSquaredError()],
)
```

Its goal is to predict the ratings as accurately as possible.

The second is the retrieval task:

```python
tfrs.tasks.Retrieval(
    metrics=tfrs.metrics.FactorizedTopK(
        candidates=movies.batch(128)
    )
)
```

As before, this task's goal is to predict which movies the user will or will not watch.

### Putting it together

We put it all together in a model class.

The new component here is that - since we have two tasks and two losses - we need to decide on how important each loss is. We can do this by giving each of the losses a weight, and treating these weights as hyperparameters. If we assign a large loss weight to the rating task, our model is going to focus on predicting ratings (but still use some information from the retrieval task); if we assign a large loss weight to the retrieval task, it will focus on retrieval instead.

In [7]:
class MovielensModel(tfrs.models.Model):

  def __init__(self, rating_weight: float, retrieval_weight: float) -> None:
    # We take the loss weights in the constructor: this allows us to instantiate
    # several model objects with different loss weights.

    super().__init__()

    embedding_dimension = 32

    # User and movie models.
    self.movie_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.experimental.preprocessing.StringLookup(
        vocabulary=unique_movie_titles, mask_token=None),
      tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
    ])
    self.user_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.experimental.preprocessing.StringLookup(
        vocabulary=unique_user_ids, mask_token=None),
      tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
    ])

    # A small model to take in user and movie embeddings and predict ratings.
    # We can make this as complicated as we want as long as we output a scalar
    # as our prediction.
    self.rating_model = tf.keras.Sequential([
        tf.keras.layers.Dense(256, activation="relu"),
        tf.keras.layers.Dense(128, activation="relu"),
        tf.keras.layers.Dense(1),
    ])

    # The tasks.
    self.rating_task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
        loss=tf.keras.losses.MeanSquaredError(),
        metrics=[tf.keras.metrics.RootMeanSquaredError()],
    )
    self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
        metrics=tfrs.metrics.FactorizedTopK(
            candidates=movies.batch(128).map(self.movie_model)
        )
    )

    # The loss weights.
    self.rating_weight = rating_weight
    self.retrieval_weight = retrieval_weight

  def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
    # We pick out the user features and pass them into the user model.
    user_embeddings = self.user_model(features["user_id"])
    # And pick out the movie features and pass them into the movie model.
    movie_embeddings = self.movie_model(features["movie_title"])
    
    return (
        user_embeddings,
        movie_embeddings,
        # We apply the multi-layered rating model to a concatentation of
        # user and movie embeddings.
        self.rating_model(
            tf.concat([user_embeddings, movie_embeddings], axis=1)
        ),
    )

  def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

    ratings = features.pop("user_rating")

    user_embeddings, movie_embeddings, rating_predictions = self(features)

    # We compute the loss for each task.
    rating_loss = self.rating_task(
        labels=ratings,
        predictions=rating_predictions,
    )
    retrieval_loss = self.retrieval_task(user_embeddings, movie_embeddings)

    # And combine them using the loss weights.
    return (self.rating_weight * rating_loss
            + self.retrieval_weight * retrieval_loss)

### Rating-specialized model

Depending on the weights we assign, the model will encode a different balance of the tasks. Let's start with a model that only considers ratings.

In [8]:
# Here, configuring the model with losses and metrics.
# TODO 1: Here is your code.
model = MovielensModel(rating_weight=1.0, retrieval_weight=0.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [9]:
cached_train = train.shuffle(100_000).batch(8192).cache()
cached_test = test.batch(4096).cache()

In [10]:
# Training the ratings model.
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 3.6922 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0031 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0575 - loss: 13.6320 - regularization_loss: 0.0000e+00 - total_loss: 13.6320

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 3.2709 - factorized_top_k/top_1_categorical_accuracy: 5.4932e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0031 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 10.6989 - regularization_loss: 0.0000e+00 - total_loss: 10.6989

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 2.8313 - factorized_top_k/top_1_categorical_accuracy: 5.2897e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0308 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 8.0165 - regularization_loss: 0.0000e+00 - total_loss: 8.0165  

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 2.8111 - factorized_top_k/top_1_categorical_accuracy: 4.5776e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0597 - loss: 7.9023 - regularization_loss: 0.0000e+00 - total_loss: 7.9023

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 2.6035 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0582 - loss: 6.7782 - regularization_loss: 0.0000e+00 - total_loss: 6.7782

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 2.4424 - factorized_top_k/top_1_categorical_accuracy: 5.0863e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0300 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 5.9655 - regularization_loss: 0.0000e+00 - total_loss: 5.9655

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 2.3456 - factorized_top_k/top_1_categorical_accuracy: 4.7084e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0591 - loss: 5.5017 - regularization_loss: 0.0000e+00 - total_loss: 5.5017

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.2365 - factorized_top_k/top_1_categorical_accuracy: 4.5776e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 5.0019 - regularization_loss: 0.0000e+00 - total_loss: 5.0019

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.1507 - factorized_top_k/top_1_categorical_accuracy: 4.2046e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0589 - loss: 4.6257 - regularization_loss: 0.0000e+00 - total_loss: 4.6257

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.0903 - factorized_top_k/top_1_categorical_accuracy: 4.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0296 - factorized_top_k/top_100_categorical_accuracy: 0.0590 - loss: 4.2989 - regularization_loss: 0.0000e+00 - total_loss: 4.2989

10/10 [==============================] - 2s 235ms/step - root_mean_squared_error: 2.0903 - factorized_top_k/top_1_categorical_accuracy: 4.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0296 - factorized_top_k/top_100_categorical_accuracy: 0.0590 - loss: 4.0315 - regularization_loss: 0.0000e+00 - total_loss: 4.0315


Epoch 2/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 1.2016 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0066 - factorized_top_k/top_50_categorical_accuracy: 0.0291 - factorized_top_k/top_100_categorical_accuracy: 0.0580 - loss: 1.4439 - regularization_loss: 0.0000e+00 - total_loss: 1.4439

 2/10 [=====>........................] - ETA: 0s - root_mean_squared_error: 1.1857 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0029 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0590 - loss: 1.4059 - regularization_loss: 0.0000e+00 - total_loss: 1.4059

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 1.1732 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0306 - factorized_top_k/top_100_categorical_accuracy: 0.0600 - loss: 1.3763 - regularization_loss: 0.0000e+00 - total_loss: 1.3763

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1660 - factorized_top_k/top_1_categorical_accuracy: 3.3569e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0601 - loss: 1.3596 - regularization_loss: 0.0000e+00 - total_loss: 1.3596

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1633 - factorized_top_k/top_1_categorical_accuracy: 3.9063e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0588 - loss: 1.3533 - regularization_loss: 0.0000e+00 - total_loss: 1.3533

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1605 - factorized_top_k/top_1_categorical_accuracy: 4.2725e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.3467 - regularization_loss: 0.0000e+00 - total_loss: 1.3467

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1581 - factorized_top_k/top_1_categorical_accuracy: 4.1853e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0299 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 1.3411 - regularization_loss: 0.0000e+00 - total_loss: 1.3411

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1574 - factorized_top_k/top_1_categorical_accuracy: 4.4250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0591 - loss: 1.3396 - regularization_loss: 0.0000e+00 - total_loss: 1.3396

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1567 - factorized_top_k/top_1_categorical_accuracy: 4.0690e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0592 - loss: 1.3379 - regularization_loss: 0.0000e+00 - total_loss: 1.3379

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1531 - factorized_top_k/top_1_categorical_accuracy: 3.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0593 - loss: 1.3275 - regularization_loss: 0.0000e+00 - total_loss: 1.3275

10/10 [==============================] - 2s 219ms/step - root_mean_squared_error: 1.1531 - factorized_top_k/top_1_categorical_accuracy: 3.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0593 - loss: 1.3189 - regularization_loss: 0.0000e+00 - total_loss: 1.3189


Epoch 3/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 1.1343 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0062 - factorized_top_k/top_50_categorical_accuracy: 0.0289 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 1.2867 - regularization_loss: 0.0000e+00 - total_loss: 1.2867

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 1.1330 - factorized_top_k/top_1_categorical_accuracy: 4.2725e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0295 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 1.2838 - regularization_loss: 0.0000e+00 - total_loss: 1.2838

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 1.1222 - factorized_top_k/top_1_categorical_accuracy: 5.6966e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0307 - factorized_top_k/top_100_categorical_accuracy: 0.0602 - loss: 1.2593 - regularization_loss: 0.0000e+00 - total_loss: 1.2593

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1203 - factorized_top_k/top_1_categorical_accuracy: 4.2725e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0606 - loss: 1.2550 - regularization_loss: 0.0000e+00 - total_loss: 1.2550

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1199 - factorized_top_k/top_1_categorical_accuracy: 4.3945e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0296 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 1.2543 - regularization_loss: 0.0000e+00 - total_loss: 1.2543

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1206 - factorized_top_k/top_1_categorical_accuracy: 4.4759e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0305 - factorized_top_k/top_100_categorical_accuracy: 0.0604 - loss: 1.2557 - regularization_loss: 0.0000e+00 - total_loss: 1.2557

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1202 - factorized_top_k/top_1_categorical_accuracy: 4.5340e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0600 - loss: 1.2549 - regularization_loss: 0.0000e+00 - total_loss: 1.2549

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1214 - factorized_top_k/top_1_categorical_accuracy: 4.4250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 1.2575 - regularization_loss: 0.0000e+00 - total_loss: 1.2575

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1219 - factorized_top_k/top_1_categorical_accuracy: 4.2046e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0301 - factorized_top_k/top_100_categorical_accuracy: 0.0598 - loss: 1.2586 - regularization_loss: 0.0000e+00 - total_loss: 1.2586

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1198 - factorized_top_k/top_1_categorical_accuracy: 3.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0598 - loss: 1.2527 - regularization_loss: 0.0000e+00 - total_loss: 1.2527

10/10 [==============================] - 2s 219ms/step - root_mean_squared_error: 1.1198 - factorized_top_k/top_1_categorical_accuracy: 3.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0055 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0598 - loss: 1.2479 - regularization_loss: 0.0000e+00 - total_loss: 1.2479


1/5 [=====>........................] - ETA: 0s - root_mean_squared_error: 1.1173 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0032 - factorized_top_k/top_10_categorical_accuracy: 0.0042 - factorized_top_k/top_50_categorical_accuracy: 0.0271 - factorized_top_k/top_100_categorical_accuracy: 0.0603 - loss: 1.2483 - regularization_loss: 0.0000e+00 - total_loss: 1.2483

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1210 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0044 - factorized_top_k/top_50_categorical_accuracy: 0.0267 - factorized_top_k/top_100_categorical_accuracy: 0.0571 - loss: 1.2565 - regularization_loss: 0.0000e+00 - total_loss: 1.2565

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 6.5104e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0046 - factorized_top_k/top_50_categorical_accuracy: 0.0280 - factorized_top_k/top_100_categorical_accuracy: 0.0578 - loss: 1.2543 - regularization_loss: 0.0000e+00 - total_loss: 1.2543

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1156 - factorized_top_k/top_1_categorical_accuracy: 5.4932e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0050 - factorized_top_k/top_50_categorical_accuracy: 0.0294 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 1.2447 - regularization_loss: 0.0000e+00 - total_loss: 1.2447

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1130 - factorized_top_k/top_1_categorical_accuracy: 5.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0053 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 1.2380 - regularization_loss: 0.0000e+00 - total_loss: 1.2380

5/5 [==============================] - 1s 107ms/step - root_mean_squared_error: 1.1130 - factorized_top_k/top_1_categorical_accuracy: 5.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0030 - factorized_top_k/top_10_categorical_accuracy: 0.0053 - factorized_top_k/top_50_categorical_accuracy: 0.0298 - factorized_top_k/top_100_categorical_accuracy: 0.0595 - loss: 1.2336 - regularization_loss: 0.0000e+00 - total_loss: 1.2336


Retrieval top-100 accuracy: 0.060.
Ranking RMSE: 1.113.


The model does OK on predicting ratings (with an RMSE of around 1.11), but performs poorly at predicting which movies will be watched or not: its accuracy at 100 is almost 4 times worse than a model trained solely to predict watches.

### Retrieval-specialized model

Let's now try a model that focuses on retrieval only.

In [11]:
# Here, configuring the model with losses and metrics.
# TODO 2: Here is your code.
model = MovielensModel(rating_weight=0.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [12]:
# Training the retrieval model.
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 3.6975 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0032 - factorized_top_k/top_10_categorical_accuracy: 0.0061 - factorized_top_k/top_50_categorical_accuracy: 0.0288 - factorized_top_k/top_100_categorical_accuracy: 0.0581 - loss: 73817.5703 - regularization_loss: 0.0000e+00 - total_loss: 73817.5703

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 3.7038 - factorized_top_k/top_1_categorical_accuracy: 5.4932e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0046 - factorized_top_k/top_10_categorical_accuracy: 0.0087 - factorized_top_k/top_50_categorical_accuracy: 0.0372 - factorized_top_k/top_100_categorical_accuracy: 0.0731 - loss: 73818.4492 - regularization_loss: 0.0000e+00 - total_loss: 73818.4492

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 3.7059 - factorized_top_k/top_1_categorical_accuracy: 6.9173e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0050 - factorized_top_k/top_10_categorical_accuracy: 0.0100 - factorized_top_k/top_50_categorical_accuracy: 0.0444 - factorized_top_k/top_100_categorical_accuracy: 0.0821 - loss: 73819.6068 - regularization_loss: 0.0000e+00 - total_loss: 73819.6068

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7098 - factorized_top_k/top_1_categorical_accuracy: 8.5449e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0057 - factorized_top_k/top_10_categorical_accuracy: 0.0114 - factorized_top_k/top_50_categorical_accuracy: 0.0515 - factorized_top_k/top_100_categorical_accuracy: 0.0932 - loss: 73815.9902 - regularization_loss: 0.0000e+00 - total_loss: 73815.9902

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.7128 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0064 - factorized_top_k/top_10_categorical_accuracy: 0.0129 - factorized_top_k/top_50_categorical_accuracy: 0.0601 - factorized_top_k/top_100_categorical_accuracy: 0.1093 - loss: 73799.9422 - regularization_loss: 0.0000e+00 - total_loss: 73799.9422    

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 3.7169 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0075 - factorized_top_k/top_10_categorical_accuracy: 0.0153 - factorized_top_k/top_50_categorical_accuracy: 0.0726 - factorized_top_k/top_100_categorical_accuracy: 0.1289 - loss: 73756.4193 - regularization_loss: 0.0000e+00 - total_loss: 73756.4193

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7211 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0176 - factorized_top_k/top_50_categorical_accuracy: 0.0830 - factorized_top_k/top_100_categorical_accuracy: 0.1471 - loss: 73678.2143 - regularization_loss: 0.0000e+00 - total_loss: 73678.2143

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7208 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0098 - factorized_top_k/top_10_categorical_accuracy: 0.0197 - factorized_top_k/top_50_categorical_accuracy: 0.0928 - factorized_top_k/top_100_categorical_accuracy: 0.1633 - loss: 73566.4336 - regularization_loss: 0.0000e+00 - total_loss: 73566.4336

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7217 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0105 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1000 - factorized_top_k/top_100_categorical_accuracy: 0.1755 - loss: 73441.5208 - regularization_loss: 0.0000e+00 - total_loss: 73441.5208

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7237 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0101 - factorized_top_k/top_10_categorical_accuracy: 0.0208 - factorized_top_k/top_50_categorical_accuracy: 0.1012 - factorized_top_k/top_100_categorical_accuracy: 0.1795 - loss: 71448.6000 - regularization_loss: 0.0000e+00 - total_loss: 71448.6000

10/10 [==============================] - 2s 219ms/step - root_mean_squared_error: 3.7237 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0101 - factorized_top_k/top_10_categorical_accuracy: 0.0208 - factorized_top_k/top_50_categorical_accuracy: 0.1012 - factorized_top_k/top_100_categorical_accuracy: 0.1795 - loss: 69818.0284 - regularization_loss: 0.0000e+00 - total_loss: 69818.0284


Epoch 2/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 3.7272 - factorized_top_k/top_1_categorical_accuracy: 0.0024 - factorized_top_k/top_5_categorical_accuracy: 0.0156 - factorized_top_k/top_10_categorical_accuracy: 0.0354 - factorized_top_k/top_50_categorical_accuracy: 0.1642 - factorized_top_k/top_100_categorical_accuracy: 0.2831 - loss: 71781.0859 - regularization_loss: 0.0000e+00 - total_loss: 71781.0859

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 3.7329 - factorized_top_k/top_1_categorical_accuracy: 0.0026 - factorized_top_k/top_5_categorical_accuracy: 0.0167 - factorized_top_k/top_10_categorical_accuracy: 0.0360 - factorized_top_k/top_50_categorical_accuracy: 0.1637 - factorized_top_k/top_100_categorical_accuracy: 0.2853 - loss: 71630.4102 - regularization_loss: 0.0000e+00 - total_loss: 71630.4102

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 3.7344 - factorized_top_k/top_1_categorical_accuracy: 0.0026 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0368 - factorized_top_k/top_50_categorical_accuracy: 0.1649 - factorized_top_k/top_100_categorical_accuracy: 0.2878 - loss: 71485.6198 - regularization_loss: 0.0000e+00 - total_loss: 71485.6198

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7380 - factorized_top_k/top_1_categorical_accuracy: 0.0027 - factorized_top_k/top_5_categorical_accuracy: 0.0186 - factorized_top_k/top_10_categorical_accuracy: 0.0371 - factorized_top_k/top_50_categorical_accuracy: 0.1677 - factorized_top_k/top_100_categorical_accuracy: 0.2910 - loss: 71338.7930 - regularization_loss: 0.0000e+00 - total_loss: 71338.7930

 5/10 [==============>...............] - ETA: 0s - root_mean_squared_error: 3.7409 - factorized_top_k/top_1_categorical_accuracy: 0.0027 - factorized_top_k/top_5_categorical_accuracy: 0.0186 - factorized_top_k/top_10_categorical_accuracy: 0.0380 - factorized_top_k/top_50_categorical_accuracy: 0.1677 - factorized_top_k/top_100_categorical_accuracy: 0.2901 - loss: 71245.5312 - regularization_loss: 0.0000e+00 - total_loss: 71245.5312

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 3.7447 - factorized_top_k/top_1_categorical_accuracy: 0.0028 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0385 - factorized_top_k/top_50_categorical_accuracy: 0.1686 - factorized_top_k/top_100_categorical_accuracy: 0.2910 - loss: 71141.9036 - regularization_loss: 0.0000e+00 - total_loss: 71141.9036

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7484 - factorized_top_k/top_1_categorical_accuracy: 0.0028 - factorized_top_k/top_5_categorical_accuracy: 0.0191 - factorized_top_k/top_10_categorical_accuracy: 0.0390 - factorized_top_k/top_50_categorical_accuracy: 0.1692 - factorized_top_k/top_100_categorical_accuracy: 0.2916 - loss: 71057.7790 - regularization_loss: 0.0000e+00 - total_loss: 71057.7790

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7475 - factorized_top_k/top_1_categorical_accuracy: 0.0029 - factorized_top_k/top_5_categorical_accuracy: 0.0193 - factorized_top_k/top_10_categorical_accuracy: 0.0391 - factorized_top_k/top_50_categorical_accuracy: 0.1695 - factorized_top_k/top_100_categorical_accuracy: 0.2928 - loss: 70959.7207 - regularization_loss: 0.0000e+00 - total_loss: 70959.7207

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7479 - factorized_top_k/top_1_categorical_accuracy: 0.0029 - factorized_top_k/top_5_categorical_accuracy: 0.0192 - factorized_top_k/top_10_categorical_accuracy: 0.0389 - factorized_top_k/top_50_categorical_accuracy: 0.1696 - factorized_top_k/top_100_categorical_accuracy: 0.2933 - loss: 70895.4774 - regularization_loss: 0.0000e+00 - total_loss: 70895.4774

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7495 - factorized_top_k/top_1_categorical_accuracy: 0.0028 - factorized_top_k/top_5_categorical_accuracy: 0.0189 - factorized_top_k/top_10_categorical_accuracy: 0.0385 - factorized_top_k/top_50_categorical_accuracy: 0.1691 - factorized_top_k/top_100_categorical_accuracy: 0.2929 - loss: 69013.2730 - regularization_loss: 0.0000e+00 - total_loss: 69013.2730

10/10 [==============================] - 2s 217ms/step - root_mean_squared_error: 3.7495 - factorized_top_k/top_1_categorical_accuracy: 0.0028 - factorized_top_k/top_5_categorical_accuracy: 0.0189 - factorized_top_k/top_10_categorical_accuracy: 0.0385 - factorized_top_k/top_50_categorical_accuracy: 0.1691 - factorized_top_k/top_100_categorical_accuracy: 0.2929 - loss: 67473.2876 - regularization_loss: 0.0000e+00 - total_loss: 67473.2876


Epoch 3/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 3.7464 - factorized_top_k/top_1_categorical_accuracy: 0.0028 - factorized_top_k/top_5_categorical_accuracy: 0.0216 - factorized_top_k/top_10_categorical_accuracy: 0.0432 - factorized_top_k/top_50_categorical_accuracy: 0.1786 - factorized_top_k/top_100_categorical_accuracy: 0.3026 - loss: 70042.7031 - regularization_loss: 0.0000e+00 - total_loss: 70042.7031

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 3.7518 - factorized_top_k/top_1_categorical_accuracy: 0.0026 - factorized_top_k/top_5_categorical_accuracy: 0.0205 - factorized_top_k/top_10_categorical_accuracy: 0.0430 - factorized_top_k/top_50_categorical_accuracy: 0.1790 - factorized_top_k/top_100_categorical_accuracy: 0.3050 - loss: 69991.5781 - regularization_loss: 0.0000e+00 - total_loss: 69991.5781

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 3.7529 - factorized_top_k/top_1_categorical_accuracy: 0.0031 - factorized_top_k/top_5_categorical_accuracy: 0.0210 - factorized_top_k/top_10_categorical_accuracy: 0.0446 - factorized_top_k/top_50_categorical_accuracy: 0.1833 - factorized_top_k/top_100_categorical_accuracy: 0.3089 - loss: 69916.0781 - regularization_loss: 0.0000e+00 - total_loss: 69916.0781

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7561 - factorized_top_k/top_1_categorical_accuracy: 0.0033 - factorized_top_k/top_5_categorical_accuracy: 0.0224 - factorized_top_k/top_10_categorical_accuracy: 0.0463 - factorized_top_k/top_50_categorical_accuracy: 0.1863 - factorized_top_k/top_100_categorical_accuracy: 0.3120 - loss: 69835.9551 - regularization_loss: 0.0000e+00 - total_loss: 69835.9551

 5/10 [==============>...............] - ETA: 0s - root_mean_squared_error: 3.7584 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0225 - factorized_top_k/top_10_categorical_accuracy: 0.0464 - factorized_top_k/top_50_categorical_accuracy: 0.1867 - factorized_top_k/top_100_categorical_accuracy: 0.3125 - loss: 69798.6469 - regularization_loss: 0.0000e+00 - total_loss: 69798.6469

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 3.7618 - factorized_top_k/top_1_categorical_accuracy: 0.0033 - factorized_top_k/top_5_categorical_accuracy: 0.0227 - factorized_top_k/top_10_categorical_accuracy: 0.0468 - factorized_top_k/top_50_categorical_accuracy: 0.1880 - factorized_top_k/top_100_categorical_accuracy: 0.3145 - loss: 69744.9076 - regularization_loss: 0.0000e+00 - total_loss: 69744.9076

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0228 - factorized_top_k/top_10_categorical_accuracy: 0.0463 - factorized_top_k/top_50_categorical_accuracy: 0.1881 - factorized_top_k/top_100_categorical_accuracy: 0.3142 - loss: 69717.8850 - regularization_loss: 0.0000e+00 - total_loss: 69717.8850

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7635 - factorized_top_k/top_1_categorical_accuracy: 0.0034 - factorized_top_k/top_5_categorical_accuracy: 0.0229 - factorized_top_k/top_10_categorical_accuracy: 0.0461 - factorized_top_k/top_50_categorical_accuracy: 0.1881 - factorized_top_k/top_100_categorical_accuracy: 0.3155 - loss: 69664.0771 - regularization_loss: 0.0000e+00 - total_loss: 69664.0771

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.7635 - factorized_top_k/top_1_categorical_accuracy: 0.0034 - factorized_top_k/top_5_categorical_accuracy: 0.0227 - factorized_top_k/top_10_categorical_accuracy: 0.0461 - factorized_top_k/top_50_categorical_accuracy: 0.1880 - factorized_top_k/top_100_categorical_accuracy: 0.3154 - loss: 69644.1215 - regularization_loss: 0.0000e+00 - total_loss: 69644.1215

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0223 - factorized_top_k/top_10_categorical_accuracy: 0.0456 - factorized_top_k/top_50_categorical_accuracy: 0.1879 - factorized_top_k/top_100_categorical_accuracy: 0.3148 - loss: 67820.9430 - regularization_loss: 0.0000e+00 - total_loss: 67820.9430

10/10 [==============================] - 2s 218ms/step - root_mean_squared_error: 3.7648 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0223 - factorized_top_k/top_10_categorical_accuracy: 0.0456 - factorized_top_k/top_50_categorical_accuracy: 0.1879 - factorized_top_k/top_100_categorical_accuracy: 0.3148 - loss: 66329.2514 - regularization_loss: 0.0000e+00 - total_loss: 66329.2514


1/5 [=====>........................] - ETA: 0s - root_mean_squared_error: 3.7755 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0115 - factorized_top_k/top_10_categorical_accuracy: 0.0254 - factorized_top_k/top_50_categorical_accuracy: 0.1272 - factorized_top_k/top_100_categorical_accuracy: 0.2383 - loss: 32458.2539 - regularization_loss: 0.0000e+00 - total_loss: 32458.2539

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 3.7762 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0095 - factorized_top_k/top_10_categorical_accuracy: 0.0216 - factorized_top_k/top_50_categorical_accuracy: 0.1244 - factorized_top_k/top_100_categorical_accuracy: 0.2358 - loss: 32508.5273 - regularization_loss: 0.0000e+00 - total_loss: 32508.5273

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 3.7744 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0090 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1253 - factorized_top_k/top_100_categorical_accuracy: 0.2347 - loss: 32516.4388 - regularization_loss: 0.0000e+00 - total_loss: 32516.4388

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.7732 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0093 - factorized_top_k/top_10_categorical_accuracy: 0.0217 - factorized_top_k/top_50_categorical_accuracy: 0.1246 - factorized_top_k/top_100_categorical_accuracy: 0.2359 - loss: 32495.7471 - regularization_loss: 0.0000e+00 - total_loss: 32495.7471    

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 3.7730 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0098 - factorized_top_k/top_10_categorical_accuracy: 0.0218 - factorized_top_k/top_50_categorical_accuracy: 0.1253 - factorized_top_k/top_100_categorical_accuracy: 0.2353 - loss: 31649.3406 - regularization_loss: 0.0000e+00 - total_loss: 31649.3406

5/5 [==============================] - 1s 108ms/step - root_mean_squared_error: 3.7730 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0098 - factorized_top_k/top_10_categorical_accuracy: 0.0218 - factorized_top_k/top_50_categorical_accuracy: 0.1253 - factorized_top_k/top_100_categorical_accuracy: 0.2353 - loss: 31085.0697 - regularization_loss: 0.0000e+00 - total_loss: 31085.0697


Retrieval top-100 accuracy: 0.235.
Ranking RMSE: 3.773.


We get the opposite result: a model that does well on retrieval, but poorly on predicting ratings.

### Joint model

Let's now train a model that assigns positive weights to both tasks.

In [13]:
# Here, configuring the model with losses and metrics.
# TODO 3: Here is your code.
model = MovielensModel(rating_weight=1.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [14]:
# Training the joint model.
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 3.6615 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0320 - factorized_top_k/top_100_categorical_accuracy: 0.0636 - loss: 73830.4453 - regularization_loss: 0.0000e+00 - total_loss: 73830.4453

 2/10 [=====>........................] - ETA: 0s - root_mean_squared_error: 3.1084 - factorized_top_k/top_1_categorical_accuracy: 7.9346e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0049 - factorized_top_k/top_10_categorical_accuracy: 0.0092 - factorized_top_k/top_50_categorical_accuracy: 0.0403 - factorized_top_k/top_100_categorical_accuracy: 0.0735 - loss: 73829.0508 - regularization_loss: 0.0000e+00 - total_loss: 73829.0508

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 3.0812 - factorized_top_k/top_1_categorical_accuracy: 8.1380e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0052 - factorized_top_k/top_10_categorical_accuracy: 0.0101 - factorized_top_k/top_50_categorical_accuracy: 0.0459 - factorized_top_k/top_100_categorical_accuracy: 0.0830 - loss: 73829.5286 - regularization_loss: 0.0000e+00 - total_loss: 73829.5286

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.1885 - factorized_top_k/top_1_categorical_accuracy: 8.5449e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0058 - factorized_top_k/top_10_categorical_accuracy: 0.0114 - factorized_top_k/top_50_categorical_accuracy: 0.0537 - factorized_top_k/top_100_categorical_accuracy: 0.0958 - loss: 73825.9199 - regularization_loss: 0.0000e+00 - total_loss: 73825.9199

 5/10 [==============>...............] - ETA: 0s - root_mean_squared_error: 3.1477 - factorized_top_k/top_1_categorical_accuracy: 8.7891e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0064 - factorized_top_k/top_10_categorical_accuracy: 0.0133 - factorized_top_k/top_50_categorical_accuracy: 0.0629 - factorized_top_k/top_100_categorical_accuracy: 0.1118 - loss: 73809.7203 - regularization_loss: 0.0000e+00 - total_loss: 73809.7203

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 3.0200 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0072 - factorized_top_k/top_10_categorical_accuracy: 0.0152 - factorized_top_k/top_50_categorical_accuracy: 0.0741 - factorized_top_k/top_100_categorical_accuracy: 0.1317 - loss: 73765.6224 - regularization_loss: 0.0000e+00 - total_loss: 73765.6224

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 2.8407 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0177 - factorized_top_k/top_50_categorical_accuracy: 0.0844 - factorized_top_k/top_100_categorical_accuracy: 0.1483 - loss: 73683.8605 - regularization_loss: 0.0000e+00 - total_loss: 73683.8605    

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.6953 - factorized_top_k/top_1_categorical_accuracy: 0.0014 - factorized_top_k/top_5_categorical_accuracy: 0.0098 - factorized_top_k/top_10_categorical_accuracy: 0.0197 - factorized_top_k/top_50_categorical_accuracy: 0.0928 - factorized_top_k/top_100_categorical_accuracy: 0.1633 - loss: 73568.7500 - regularization_loss: 0.0000e+00 - total_loss: 73568.7500

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.5823 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0105 - factorized_top_k/top_10_categorical_accuracy: 0.0211 - factorized_top_k/top_50_categorical_accuracy: 0.0996 - factorized_top_k/top_100_categorical_accuracy: 0.1745 - loss: 73439.2422 - regularization_loss: 0.0000e+00 - total_loss: 73439.2422

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.5007 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0100 - factorized_top_k/top_10_categorical_accuracy: 0.0207 - factorized_top_k/top_50_categorical_accuracy: 0.1009 - factorized_top_k/top_100_categorical_accuracy: 0.1788 - loss: 71444.1621 - regularization_loss: 0.0000e+00 - total_loss: 71444.1621

10/10 [==============================] - 2s 216ms/step - root_mean_squared_error: 2.5007 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0100 - factorized_top_k/top_10_categorical_accuracy: 0.0207 - factorized_top_k/top_50_categorical_accuracy: 0.1009 - factorized_top_k/top_100_categorical_accuracy: 0.1788 - loss: 69811.8239 - regularization_loss: 0.0000e+00 - total_loss: 69811.8239


Epoch 2/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 1.2498 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0160 - factorized_top_k/top_10_categorical_accuracy: 0.0364 - factorized_top_k/top_50_categorical_accuracy: 0.1609 - factorized_top_k/top_100_categorical_accuracy: 0.2773 - loss: 71774.2812 - regularization_loss: 0.0000e+00 - total_loss: 71774.2812

 2/10 [=====>........................] - ETA: 0s - root_mean_squared_error: 1.2114 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0168 - factorized_top_k/top_10_categorical_accuracy: 0.0350 - factorized_top_k/top_50_categorical_accuracy: 0.1556 - factorized_top_k/top_100_categorical_accuracy: 0.2766 - loss: 71638.0938 - regularization_loss: 0.0000e+00 - total_loss: 71638.0938

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 1.2093 - factorized_top_k/top_1_categorical_accuracy: 0.0022 - factorized_top_k/top_5_categorical_accuracy: 0.0168 - factorized_top_k/top_10_categorical_accuracy: 0.0353 - factorized_top_k/top_50_categorical_accuracy: 0.1571 - factorized_top_k/top_100_categorical_accuracy: 0.2802 - loss: 71496.3021 - regularization_loss: 0.0000e+00 - total_loss: 71496.3021

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.2012 - factorized_top_k/top_1_categorical_accuracy: 0.0025 - factorized_top_k/top_5_categorical_accuracy: 0.0173 - factorized_top_k/top_10_categorical_accuracy: 0.0360 - factorized_top_k/top_50_categorical_accuracy: 0.1609 - factorized_top_k/top_100_categorical_accuracy: 0.2838 - loss: 71351.8613 - regularization_loss: 0.0000e+00 - total_loss: 71351.8613

 5/10 [==============>...............] - ETA: 0s - root_mean_squared_error: 1.2119 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0178 - factorized_top_k/top_10_categorical_accuracy: 0.0365 - factorized_top_k/top_50_categorical_accuracy: 0.1621 - factorized_top_k/top_100_categorical_accuracy: 0.2837 - loss: 71257.9094 - regularization_loss: 0.0000e+00 - total_loss: 71257.9094

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 1.2056 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0367 - factorized_top_k/top_50_categorical_accuracy: 0.1630 - factorized_top_k/top_100_categorical_accuracy: 0.2857 - loss: 71155.9010 - regularization_loss: 0.0000e+00 - total_loss: 71155.9010

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.2128 - factorized_top_k/top_1_categorical_accuracy: 0.0024 - factorized_top_k/top_5_categorical_accuracy: 0.0186 - factorized_top_k/top_10_categorical_accuracy: 0.0373 - factorized_top_k/top_50_categorical_accuracy: 0.1638 - factorized_top_k/top_100_categorical_accuracy: 0.2859 - loss: 71071.9475 - regularization_loss: 0.0000e+00 - total_loss: 71071.9475

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.2100 - factorized_top_k/top_1_categorical_accuracy: 0.0024 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0377 - factorized_top_k/top_50_categorical_accuracy: 0.1644 - factorized_top_k/top_100_categorical_accuracy: 0.2874 - loss: 70974.7520 - regularization_loss: 0.0000e+00 - total_loss: 70974.7520

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.2166 - factorized_top_k/top_1_categorical_accuracy: 0.0025 - factorized_top_k/top_5_categorical_accuracy: 0.0186 - factorized_top_k/top_10_categorical_accuracy: 0.0371 - factorized_top_k/top_50_categorical_accuracy: 0.1643 - factorized_top_k/top_100_categorical_accuracy: 0.2878 - loss: 70909.9314 - regularization_loss: 0.0000e+00 - total_loss: 70909.9314

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.2097 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0184 - factorized_top_k/top_10_categorical_accuracy: 0.0367 - factorized_top_k/top_50_categorical_accuracy: 0.1638 - factorized_top_k/top_100_categorical_accuracy: 0.2875 - loss: 69024.1691 - regularization_loss: 0.0000e+00 - total_loss: 69024.1691

10/10 [==============================] - 2s 217ms/step - root_mean_squared_error: 1.2097 - factorized_top_k/top_1_categorical_accuracy: 0.0023 - factorized_top_k/top_5_categorical_accuracy: 0.0184 - factorized_top_k/top_10_categorical_accuracy: 0.0367 - factorized_top_k/top_50_categorical_accuracy: 0.1638 - factorized_top_k/top_100_categorical_accuracy: 0.2875 - loss: 67481.2727 - regularization_loss: 0.0000e+00 - total_loss: 67481.2727


Epoch 3/3


 1/10 [==>...........................] - ETA: 0s - root_mean_squared_error: 1.1662 - factorized_top_k/top_1_categorical_accuracy: 0.0024 - factorized_top_k/top_5_categorical_accuracy: 0.0187 - factorized_top_k/top_10_categorical_accuracy: 0.0428 - factorized_top_k/top_50_categorical_accuracy: 0.1776 - factorized_top_k/top_100_categorical_accuracy: 0.3024 - loss: 70021.1250 - regularization_loss: 0.0000e+00 - total_loss: 70021.1250

 2/10 [=====>........................] - ETA: 1s - root_mean_squared_error: 1.1380 - factorized_top_k/top_1_categorical_accuracy: 0.0030 - factorized_top_k/top_5_categorical_accuracy: 0.0196 - factorized_top_k/top_10_categorical_accuracy: 0.0419 - factorized_top_k/top_50_categorical_accuracy: 0.1753 - factorized_top_k/top_100_categorical_accuracy: 0.3033 - loss: 69972.3867 - regularization_loss: 0.0000e+00 - total_loss: 69972.3867

 3/10 [========>.....................] - ETA: 1s - root_mean_squared_error: 1.1280 - factorized_top_k/top_1_categorical_accuracy: 0.0029 - factorized_top_k/top_5_categorical_accuracy: 0.0207 - factorized_top_k/top_10_categorical_accuracy: 0.0436 - factorized_top_k/top_50_categorical_accuracy: 0.1808 - factorized_top_k/top_100_categorical_accuracy: 0.3079 - loss: 69890.3594 - regularization_loss: 0.0000e+00 - total_loss: 69890.3594

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1207 - factorized_top_k/top_1_categorical_accuracy: 0.0033 - factorized_top_k/top_5_categorical_accuracy: 0.0220 - factorized_top_k/top_10_categorical_accuracy: 0.0448 - factorized_top_k/top_50_categorical_accuracy: 0.1843 - factorized_top_k/top_100_categorical_accuracy: 0.3123 - loss: 69806.6328 - regularization_loss: 0.0000e+00 - total_loss: 69806.6328

 5/10 [==============>...............] - ETA: 0s - root_mean_squared_error: 1.1203 - factorized_top_k/top_1_categorical_accuracy: 0.0033 - factorized_top_k/top_5_categorical_accuracy: 0.0224 - factorized_top_k/top_10_categorical_accuracy: 0.0451 - factorized_top_k/top_50_categorical_accuracy: 0.1847 - factorized_top_k/top_100_categorical_accuracy: 0.3122 - loss: 69769.5031 - regularization_loss: 0.0000e+00 - total_loss: 69769.5031

 6/10 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1157 - factorized_top_k/top_1_categorical_accuracy: 0.0034 - factorized_top_k/top_5_categorical_accuracy: 0.0228 - factorized_top_k/top_10_categorical_accuracy: 0.0455 - factorized_top_k/top_50_categorical_accuracy: 0.1865 - factorized_top_k/top_100_categorical_accuracy: 0.3146 - loss: 69715.3125 - regularization_loss: 0.0000e+00 - total_loss: 69715.3125

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1170 - factorized_top_k/top_1_categorical_accuracy: 0.0035 - factorized_top_k/top_5_categorical_accuracy: 0.0231 - factorized_top_k/top_10_categorical_accuracy: 0.0452 - factorized_top_k/top_50_categorical_accuracy: 0.1859 - factorized_top_k/top_100_categorical_accuracy: 0.3147 - loss: 69686.5223 - regularization_loss: 0.0000e+00 - total_loss: 69686.5223

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1169 - factorized_top_k/top_1_categorical_accuracy: 0.0035 - factorized_top_k/top_5_categorical_accuracy: 0.0231 - factorized_top_k/top_10_categorical_accuracy: 0.0454 - factorized_top_k/top_50_categorical_accuracy: 0.1864 - factorized_top_k/top_100_categorical_accuracy: 0.3157 - loss: 69635.3418 - regularization_loss: 0.0000e+00 - total_loss: 69635.3418

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1223 - factorized_top_k/top_1_categorical_accuracy: 0.0034 - factorized_top_k/top_5_categorical_accuracy: 0.0227 - factorized_top_k/top_10_categorical_accuracy: 0.0449 - factorized_top_k/top_50_categorical_accuracy: 0.1865 - factorized_top_k/top_100_categorical_accuracy: 0.3154 - loss: 69617.1111 - regularization_loss: 0.0000e+00 - total_loss: 69617.1111

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0223 - factorized_top_k/top_10_categorical_accuracy: 0.0443 - factorized_top_k/top_50_categorical_accuracy: 0.1862 - factorized_top_k/top_100_categorical_accuracy: 0.3150 - loss: 67791.5617 - regularization_loss: 0.0000e+00 - total_loss: 67791.5617

10/10 [==============================] - 2s 216ms/step - root_mean_squared_error: 1.1200 - factorized_top_k/top_1_categorical_accuracy: 0.0032 - factorized_top_k/top_5_categorical_accuracy: 0.0223 - factorized_top_k/top_10_categorical_accuracy: 0.0443 - factorized_top_k/top_50_categorical_accuracy: 0.1862 - factorized_top_k/top_100_categorical_accuracy: 0.3150 - loss: 66297.9304 - regularization_loss: 0.0000e+00 - total_loss: 66297.9304


1/5 [=====>........................] - ETA: 0s - root_mean_squared_error: 1.1393 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0222 - factorized_top_k/top_50_categorical_accuracy: 0.1243 - factorized_top_k/top_100_categorical_accuracy: 0.2341 - loss: 32468.5820 - regularization_loss: 0.0000e+00 - total_loss: 32468.5820

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1398 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0089 - factorized_top_k/top_10_categorical_accuracy: 0.0215 - factorized_top_k/top_50_categorical_accuracy: 0.1216 - factorized_top_k/top_100_categorical_accuracy: 0.2313 - loss: 32500.5615 - regularization_loss: 0.0000e+00 - total_loss: 32500.5615

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1364 - factorized_top_k/top_1_categorical_accuracy: 8.1380e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0222 - factorized_top_k/top_50_categorical_accuracy: 0.1236 - factorized_top_k/top_100_categorical_accuracy: 0.2321 - loss: 32499.3945 - regularization_loss: 0.0000e+00 - total_loss: 32499.3945

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1337 - factorized_top_k/top_1_categorical_accuracy: 9.1553e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0085 - factorized_top_k/top_10_categorical_accuracy: 0.0214 - factorized_top_k/top_50_categorical_accuracy: 0.1237 - factorized_top_k/top_100_categorical_accuracy: 0.2335 - loss: 32477.6860 - regularization_loss: 0.0000e+00 - total_loss: 32477.6860

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1312 - factorized_top_k/top_1_categorical_accuracy: 8.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0219 - factorized_top_k/top_50_categorical_accuracy: 0.1248 - factorized_top_k/top_100_categorical_accuracy: 0.2344 - loss: 31628.7672 - regularization_loss: 0.0000e+00 - total_loss: 31628.7672

5/5 [==============================] - 1s 111ms/step - root_mean_squared_error: 1.1312 - factorized_top_k/top_1_categorical_accuracy: 8.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0219 - factorized_top_k/top_50_categorical_accuracy: 0.1248 - factorized_top_k/top_100_categorical_accuracy: 0.2344 - loss: 31062.8213 - regularization_loss: 0.0000e+00 - total_loss: 31062.8213


Retrieval top-100 accuracy: 0.234.
Ranking RMSE: 1.131.


The result is a model that performs roughly as well on both tasks as each specialized model. 

While the results here do not show a clear accuracy benefit from a joint model in this case, multi-task learning is in general an extremely useful tool. We can expect better results when we can transfer knowledge from a data-abundant task (such as clicks) to a closely related data-sparse task (such as purchases).